Big UX and Performance Improvement Updates

404Wolf opened the pull request:
Sorry for a kinda-large PR, I got carried away a bit, but this packs a lot of nice improvements to vt. I'll try to make them smaller going forward :)

- A big reorganization refactor for commands, moving them all to their own files
  - This will make adding cli command tests easier. We can make foobar_test.ts files for each command.
- Walk up until we find the .vt dir, like git, up to 100 dirs up.
- Better error handling for commands. Throw `Deno.errors.<>` errors, and re-catch them in Commands.
- Add a `displayStatusChanges` that takes a `StatusResult`, and:
  - Print status for `push` and `pull`, showing what was pushed and pulled
  - Print rolling logs for `watch` with status info for what was pushed
- If there is no env var set, guide them to the website using `npm:open`, and ask them for the key to set it one-off, and let them know to export it or set in `.env` (I'm working in https://github.com/val-town/vt/pull/25 on a persistent config store, this is a temporary/start to a larger improvement here).
- Don't re-run `status` (which is super expensive), instead re-use the result. This was a bottleneck for commands like `pull` where we were checking dirty-ness (which called status()), and then afterwards running pull(), which called it again.
- Cleaning up various other logic, like taking advantage of stainless async generator pagination in git logic

maxmcd commented:
Wow, excellent. Few things:

Seems to complain now without a .vtignore
```
$ deno run -A ../vt/vt.ts pull
⠋ Pulling latest changes...The file /Users/maxm/go/src/github.com/val-town/apricot/.vtignore was not found
Error loading /Users/maxm/go/src/github.com/val-town/apricot/.vtignore, skipping
Nothing new to pull, everything is up to date.
```

It might be nice to report the type of val we create if we do create a val. Might also be a good time to print the http url as well. 
```
--- watching for changes (@11:23:27 AM) ---

New changes detected
A main.http.tsx

Pushed:
  1 created
```

Filed an issue related to pushing something: https://github.com/val-town/vt/issues/37 Although prob not entirely related to this PR, but just fyi I hit that while testing this branch.

Related to that and some other debugging: I wonder if we'd be better off logging changes/updates line by line? I have not looked at the internals, so maybe that is annoying, but all cli actions printing the actions they're taking (deleting, creating, updating), might help bring a lot of clarity? The changes here are more of a git-style summary, but maybe it is helpful for user comprehension and bug reports/debugging to switch to `Pushed new http val main.http.tsx` as it is happening?

maxmcd commented on vt.ts:
```diff
@@ -1,48 +1,60 @@
+  console.log(colors.green("API key set successfully!"));
+  console.log();
+  console.log(
+    "Note: This key is only set for the current session. To make it permanent, add it to your environment with:" +
+      "\n  `export VAL_TOWN_API_KEY=" + apiKey + "`" +
+      "\nor add it to your .env file.",
```
I'm not too sold on the utility here because it will not persist into the next cli command. Maybe after opening the url we just print instructions for how to set it in the environment and exit?

404Wolf replied:
This is more of a placeholder, since there's a different Pr that exposes an API to persistently store in a xdg store config.json that persists. I agree that as-is it is not super useful

maxmcd replied:
ah cool, makes sense

maxmcd commented on vt.ts:
```diff
@@ -1,48 +1,60 @@
-  const executablePath = Deno.execPath();
-  const scriptModule = Deno.mainModule;
+  if (goToWebsite) {
+    const open = await import("npm:open");
```
why the dynamic import here?

404Wolf replied:
For `npm:open` it shouldn't be dynamic, I made the vt root.ts import dynamic because it loads sdk.ts, which will (on import) try constructing a new ValTown and load the VAL_TOWN_API_KEY. I'll change this.

404Wolf commented:
> Wow, excellent. Few things:
> 
> Seems to complain now without a .vtignore
> 
> ```
> $ deno run -A ../vt/vt.ts pull
> ⠋ Pulling latest changes...The file /Users/maxm/go/src/github.com/val-town/apricot/.vtignore was not found
> Error loading /Users/maxm/go/src/github.com/val-town/apricot/.vtignore, skipping
> Nothing new to pull, everything is up to date.
> ```
> 
> It might be nice to report the type of val we create if we do create a val. Might also be a good time to print the http url as well.
> 
> ```
> --- watching for changes (@11:23:27 AM) ---
> 
> New changes detected
> A main.http.tsx
> 
> Pushed:
>   1 created
> ```
> 
> Filed an issue related to pushing something: #37 Although prob not entirely related to this PR, but just fyi I hit that while testing this branch.
> 
> Related to that and some other debugging: I wonder if we'd be better off logging changes/updates line by line? I have not looked at the internals, so maybe that is annoying, but all cli actions printing the actions they're taking (deleting, creating, updating), might help bring a lot of clarity? The changes here are more of a git-style summary, but maybe it is helpful for user comprehension and bug reports/debugging to switch to something like `Pushed new http val main.http.tsx` as it is happening?

I agree that reporting the val type would be nice! That's a good suggestion.
The reason it prints out the summary right now is because A: the action is atomic, so it either makes all the updates indicated in the summary, or it fails, and B: because that's how git does it. Do you think that it should print it out line by line as it does it? The issue is that if something goes wrong everything printed up to that point is void.

maxmcd commented:
Oh I would for sure take DEBUG logs though, if we could enable some kind of debug mode. But not rush to do that, and if you agree I can help add those as needed in the future.

404Wolf commented:
> Oh I would for sure take DEBUG logs though, if we could enable some kind of debug mode. But not rush to do that, and if you agree I can help add those as needed in the future.

Yes for sure! Do you suggest any ts log libraries? We should definitely add --verbose or --logfile or something, having logs will be super handy.

maxmcd commented:
@404Wolf I am not so familiar with the options. We use `pino` internally. Maybe @tmcw has opinions here?

404Wolf commented on src/vt/git/paths.ts:
```diff
@@ -66,31 +66,22 @@ async function getProjectItemType(
-  );
+  if (ignoreGlobs.length === 0) return false;
+
+  // All the libraries for this kinda suck, but this mostly works. Note that
```
I had no idea gitignores were so complicated!

tmcw replied:
The most established ignore implementation is https://github.com/kaelzhang/node-ignore - used by eslint, [globby](https://github.com/sindresorhus/globby), etc, and has the sindresorhus mark of approval. There might be other downsides of it that i'm not seeing, though!

404Wolf commented:
> Wow, excellent. Few things:
> 
> Seems to complain now without a .vtignore
> 
> ```
> $ deno run -A ../vt/vt.ts pull
> ⠋ Pulling latest changes...The file /Users/maxm/go/src/github.com/val-town/apricot/.vtignore was not found
> Error loading /Users/maxm/go/src/github.com/val-town/apricot/.vtignore, skipping
> Nothing new to pull, everything is up to date.
> ```
> 
> It might be nice to report the type of val we create if we do create a val. Might also be a good time to print the http url as well.
> 
> ```
> --- watching for changes (@11:23:27 AM) ---
> 
> New changes detected
> A main.http.tsx
> 
> Pushed:
>   1 created
> ```
> 
> Filed an issue related to pushing something: #37 Although prob not entirely related to this PR, but just fyi I hit that while testing this branch.
> 
> Related to that and some other debugging: I wonder if we'd be better off logging changes/updates line by line? I have not looked at the internals, so maybe that is annoying, but all cli actions printing the actions they're taking (deleting, creating, updating), might help bring a lot of clarity? The changes here are more of a git-style summary, but maybe it is helpful for user comprehension and bug reports/debugging to switch to something like `Pushed new http val main.http.tsx` as it is happening?

![c27c7a0d-edee-4bf5-a6b2-0cf4a2fc60b5](https://github.com/user-attachments/assets/fe4cafa3-b41a-43b8-b14b-aca57d82d7b1)

I fixed-ish the gitignore issue. It turns out that [gitignores are hard](https://git-scm.com/docs/gitignore), so I'm now [using this library](https://jsr.io/@ivanmaxlogiudice/gitignore) which works mostly, but we'll eventually want to add some coverage for.

maxmcd commented:
Great! Because of the size of this refactor and to keep things proceeding I am likely just going to poke around with the new functionality at the CLI level, but lemme know if there's anything you'd like eyes on.

404Wolf commented:
> Great! Because of the size of this refactor and to keep things proceeding I am likely just going to poke around with the new functionality at the CLI level, but lemme know if there's anything you'd like eyes on.

Sounds like a good idea! Might be a good idea to play with vtignore behavior with complexish rules, and making sure status reports seem correct for push and pull (including val types).

maxmcd commented:
Ok, this project: https://www.val.town/x/maxm/apricot/code/ddng-exp

I have the ddng-exp dir deleted locally:

```
maxm 👻 :~/go/src/github.com/val-town/apricot
$ ls -lah
drwxr-xr-x@  8 maxm  staff   256B Mar 17 13:48 .
drwxr-xr-x@ 34 maxm  staff   1.1K Mar 14 08:13 ..
-rw-r--r--@  1 maxm  staff   6.0K Mar 17 13:52 .DS_Store
drwxr-xr-x@  3 maxm  staff    96B Mar 17 13:52 .vt
-rw-r--r--@  1 maxm  staff    15B Mar 16 11:40 .vtignore
-rw-r--r--@  1 maxm  staff     0B Mar 16 11:23 foo.txt
-rw-r--r--@  1 maxm  staff    82B Mar 17 13:47 hello.http.ts
-rw-r--r--@  1 maxm  staff    37B Mar 16 11:23 hello.txt
```

I do `vt pull`

```
$ deno run -A ../vt/vt.ts pull
D (file  ) ddng-exp/README.md
D (script) ddng-exp/main.ts
D (script) ddng-exp/guest.ts
D (file  ) ddng-exp/deno.lock
D (file  ) ddng-exp/deno.json

Changes pulled:
  5 deleted
```

It says 5 deleted, but nothing has been deleted locally or in the remote.

404Wolf commented:
> Ok, this project: https://www.val.town/x/maxm/apricot/code/ddng-exp
> 
> I have the ddng-exp dir deleted locally:
> 
> ```
> maxm 👻 :~/go/src/github.com/val-town/apricot
> $ ls -lah
> drwxr-xr-x@  8 maxm  staff   256B Mar 17 13:48 .
> drwxr-xr-x@ 34 maxm  staff   1.1K Mar 14 08:13 ..
> -rw-r--r--@  1 maxm  staff   6.0K Mar 17 13:52 .DS_Store
> drwxr-xr-x@  3 maxm  staff    96B Mar 17 13:52 .vt
> -rw-r--r--@  1 maxm  staff    15B Mar 16 11:40 .vtignore
> -rw-r--r--@  1 maxm  staff     0B Mar 16 11:23 foo.txt
> -rw-r--r--@  1 maxm  staff    82B Mar 17 13:47 hello.http.ts
> -rw-r--r--@  1 maxm  staff    37B Mar 16 11:23 hello.txt
> ```
> 
> I do `vt pull`
> 
> ```
> $ deno run -A ../vt/vt.ts pull
> D (file  ) ddng-exp/README.md
> D (script) ddng-exp/main.ts
> D (script) ddng-exp/guest.ts
> D (file  ) ddng-exp/deno.lock
> D (file  ) ddng-exp/deno.json
> 
> Changes pulled:
>   5 deleted
> ```
> 
> It says 5 deleted, but nothing has been deleted locally or in the remote.

Ok, so I think this is a more fundamental issue that I feel needs to be addressed in a different PR. Basically:

- Right now we only look at how the local files compare to remote files, as if the local files were changed. So if you delete a folder with files, then it looks like locally you deleted stuff, even though when you pull you're getting stuff.
- I reverted the log for pull to not give you information for what happened. I think we need to make `status` bidirectional, which is nontrivial and should be a different PR. At a minimum, we can have `clone` and `pull` report what they did, and then report that to the user, without doing a diff first.

- I changed the internal logic of `pull`, since I think the previous logic was faulty. We are wrapping `clone`, but then correcting for files that should get deleted, while also making sure files created locally that were never pushed, along with ignored files, are not deleted. It's kinda tricky.
```ts
export function pull({
  targetDir,
  projectId,
  branchId,
  version,
  gitignoreRules,
}: {
  targetDir: string;
  projectId: string;
  branchId: string;
  version: number;
  gitignoreRules: string[];
}): Promise<void> {
  return doAtomically(
    async (tempDir) => {
      // Copy over all the files in the original dir into the temp dir
      await copy(targetDir, tempDir, {
        preserveTimestamps: true,
        overwrite: true,
      });

      // Clone all the files from the project into the temp dir
      await clone({
        targetDir: tempDir,
        projectId,
        branchId,
        version,
        gitignoreRules,
      });

      // Then delete all files that should no longer exist. So fetch all the
      // files from the server, and then look at all the files we cloned. After
      // cloning, we should not have files that no longer exist on the server,
      // so we can remove them.
      //
      // We do it this way where we copy over the current contents and then
      // clone because we want to keep files that exist locally and were never
      // pushed, along with all the ignored files.
      const files = new Set<string>();
      for await (
        const file of sdk.projects.files.list(projectId, {
          branch_id: branchId,
          version,
          recursive: true,
        })
      ) {
        files.add(file.path);
      }
      for await (const entry of walk(tempDir)) {
        const relativePath = relative(tempDir, entry.path);
        if (shouldIgnore(relativePath, gitignoreRules)) continue;
        if (entry.path === "" || entry.path === tempDir) continue;
        if (!files.has(relativePath)) {
          await Deno.remove(entry.path, { recursive: true });
        }
      }
    },
    targetDir,
    "vt_pull_",
  );
}
```

maxmcd commented:
@404Wolf ok, other than that this looks good to me. All my other testing worked well. So down to approve depending on where you get with your musings.

tmcw commented on src/cmd/git/pull.ts:
```diff
@@ -0,0 +1,26 @@
+      const statusResult = await vt.status();
+      if (!force && await vt.isDirty({ statusResult })) {
+        spinner.fail(dirtyErrorMsg("pull"));
+        return;
```
Should we be setting non-zero exit statuses here?

tmcw commented on src/cmd/git/checkout.ts:
```diff
@@ -0,0 +1,105 @@
+            "Branch name is required. Use -b to create a new branch " +
+              toListBranches,
+          );
+          return;
```
Probably want a non-zero exit status

404Wolf replied:
I agree! I've updated doWithSpinner to exit with 1 if there's an error, and changed the spinner.fails to throw errors. It feels a little clunky, but it works. Thanks for pointing this out
